tests/TestCase/Utility/HashTest.php

```diff
@@ -759,6 +761,344 @@ public function testMergeVariadic(): void
+                    'float' => 'float',
+                    'int' => 'int',
+                    'bool' => 'bool',
+                    'mixed' => 'mixed',
+                    'class' => stdClass::class,
+                ],
+            ]
+        );
+        $this->assertempty($errors);
+
+        // test all invalid basic types
+        $errors = Hash::validate(
+            ['null' => 1, 'array' => null, 'string' => 1.23, 'float' => 2, 'int' => null, 'bool' => new stdClass(), 'class' => true],
+            [
+                'fields' => [
+                    'null' => 'null',
+                    'array' => 'array',
+                    'string' => 'string',
+                    'float' => 'float',
+                    'int' => 'int',
+                    'bool' => 'bool',
+                    'class' => stdClass::class,
+                ],
+            ]
+        );
+        $this->assertSame(
+            [
+                'null' => 'Type for field does not match expected type.',
+                'array' => 'Type for field does not match expected type.',
+                'string' => 'Type for field does not match expected type.',
+                'float' => 'Type for field does not match expected type.',
+                'int' => 'Type for field does not match expected type.',
+                'bool' => 'Type for field does not match expected type.',
+                'class' => 'Type for field does not match expected type.',
+            ],
+            $errors
+        );
+
+        // test class-string which uses an array type-spec
+        $errors = Hash::validate(
+            ['has_class_string' => stdClass::class],
+            [
+                'fields' => [
+                    'has_class_string' => [
+                        'type' => ['class-string<>' => stdClass::class],
+                    ],
+                ],
+            ]
+        );
+        $this->assertEmpty($errors);
+
+        $errors = Hash::validate(
+            ['has_class_string' => Hash::class],
+            [
+                'fields' => [
+                    'has_class_string' => [
+                        'type' => ['class-string<>' => stdClass::class],
+                    ],
+                ],
+            ]
+        );
+        $this->assertSame(
+            ['has_class_string' => 'Type for field does not match expected type.'],
+            $errors
+        );
+    }
+
+    public function testValidateArrayTypes(): void
+    {
+        // test single-dimension
+        $errors = Hash::validate(
+            ['has_array' => ['int' => 1]],
+            [
+                'fields' => [
+                    'has_array' => [
+                        'type' => [
+                            'array{}' => [
+                                'fields' => ['int' => 'int'],
+                            ],
+                        ],
+                    ],
+                ],
+            ]
+        );
+        $this->assertEmpty($errors);
```

Comment:
Are there any lessons we can take from openschema's definition format? Having an interface that reuses key terms from a more broadly used specification could help build familiarity with what we're providing.

Reply:
I'm not familiar with openschema. Can you link it?

Reply:
Derp, I meant to say [open api schema](https://swagger.io/specification/)

Reply:
The key thing we're missing compared to open api and json schema is extra type information.

json schema has a union type but open api doesn't. However, they both have these schema compositions with `anyOf` that takes an array of type definitions.

I'd like to stay close to accepted php docblock and type formats where possible.